### PR TITLE
Spring security, user password encode, auth provider custom

### PR DESCRIPTION
**Pull request by nlperez11**
## Summary
- Spring Security Integration with CustomAuthenticationProvider
- Custom access rules for some endpoints
- unit testing with spring user and password
- Added password attribute
- Bean for password encoder resource
- Spring security test dependency
## Trello Card
https://trello.com/c/FlkL94E2/23-integrar-spring-security

### CODE REVIEW

File: src/main/java/com/wolox/training/security/CustomAuthenticationProvider.java [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import java.util.ArrayList;
+
+@Component
+public class CustomAuthenticationProvider implements AuthenticationProvider {
+
+    @Override
+    public Authentication authenticate(Authentication authentication)
+            throws AuthenticationException {
+
+        String name = authentication.getName();
```
Comment (martinremis) [inline src/main/java/com/wolox/training/security/CustomAuthenticationProvider.java]: You should authenticate with the database user password, not with a server level password

File: src/main/java/com/wolox/training/controller/UserController.java [hunk 2]
```diff
@@ -95,7 +107,9 @@ public User update(
             throw new UserIdMismatchException("Id doesn't not match");
         }
 
-        userRepository.findById(id).orElseThrow(() -> new UserNotFoundException("User not found"));
+        User u = userRepository.findById(id).orElseThrow(() -> new UserNotFoundException("User not found"));
+        user.setPassword(u.getPassword());
```
Comment (martinremis) [inline src/main/java/com/wolox/training/controller/UserController.java:111]: You have to encode it at updating too.

Comment (martinremis) [reply src/main/java/com/wolox/training/controller/UserController.java:111]: Maybe you could move the encryption to the domain setter

**Review by martinremis — CHANGES_REQUESTED** [review]

File: src/main/java/com/wolox/training/controller/UserController.java (hunk 2 above)
Comment (nlperez11) [reply src/main/java/com/wolox/training/controller/UserController.java:111]: The update service keep the encoded password and not permit update password in this resource, the encode process for password is alocated in create user service. This is a good work  ?

File: src/main/java/com/wolox/training/security/CustomAuthenticationProvider.java (hunk 1 above)
Comment (nlperez11) [reply src/main/java/com/wolox/training/security/CustomAuthenticationProvider.java]: Solved, i integrated the authentication service with a validation of users in database.

File: src/main/java/com/wolox/training/controller/UserController.java (hunk 2 above)
Comment (martinremis) [reply src/main/java/com/wolox/training/controller/UserController.java:111]: Ok, I get it now :stuck_out_tongue:

File: src/main/java/com/wolox/training/security/CustomAuthenticationProvider.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+    public Authentication authenticate(Authentication authentication)
+            throws AuthenticationException {
+
+        String username = authentication.getName();
+        String password = authentication.getCredentials().toString();
+        UserDetails userDetails = authService.loadUserByUsername(username);
+
+        return encoder().matches(password, userDetails.getPassword())
+                ? new UsernamePasswordAuthenticationToken(username, password, new ArrayList<>())
+                : null;
```
Comment (martinremis) [inline src/main/java/com/wolox/training/security/CustomAuthenticationProvider.java:33]: What happens here if `null` is returned? It throws an authentication error?

Comment (nlperez11) [reply src/main/java/com/wolox/training/security/CustomAuthenticationProvider.java:33]: If null is returned, the request's response is 401 Unauthorized.

**Review by martinremis — APPROVED** [review]

**Review by jeymarsonwolox — APPROVED** [review]

**Pull request merged**